value_source.go

```diff
@@ -0,0 +1,108 @@
+package cli
+
+import (
+	"fmt"
+	"os"
+	"strings"
+)
+
+// ValueSource is a source which can be used to look up a value,
+// typically for use with a cli.Flag
+type ValueSource interface {
+	fmt.Stringer
+	fmt.GoStringer
+
+	// Lookup returns the value from the source and if it was found
+	// or returns an empty string and false
+	Lookup() (string, bool)
+}
+
+// ValueSourceChain is a slice of ValueSource that allows for
+// lookup where the first ValueSource to resolve is returned
+type ValueSourceChain struct {
+	Chain []ValueSource
+}
+
+func (vsc *ValueSourceChain) String() string {
+	s := []string{}
+
+	for _, vs := range vsc.Chain {
+		s = append(s, vs.String())
+	}
+
+	return strings.Join(s, ", ")
+}
+
+func (vsc *ValueSourceChain) GoString() string {
+	s := []string{}
+
+	for _, vs := range vsc.Chain {
+		s = append(s, vs.GoString())
+	}
+
+	return fmt.Sprintf("ValueSourceChain{Chain:{%[1]s}}", strings.Join(s, ","))
+}
+
+func (vsc *ValueSourceChain) Lookup() (string, bool) {
+	s, _, ok := vsc.LookupWithSource()
+	return s, ok
+}
+
+func (vsc *ValueSourceChain) LookupWithSource() (string, ValueSource, bool) {
+	for _, src := range vsc.Chain {
+		if value, found := src.Lookup(); found {
+			return value, src, true
+		}
+	}
+
+	return "", nil, false
+}
+
+// envVarValueSource encapsulates a ValueSource from an environment variable
+type envVarValueSource struct {
```

Comment:
```
type envVarValueSource string
```

Reply:
Roughly the same comments about literals and func receivers for simple types as with slice types, and also future-proofing. I did this on purpose ❤️